src/smolagents/logger.py

```diff
@@ -0,0 +1,284 @@
+import json
+from dataclasses import asdict, dataclass
+from enum import IntEnum
+from typing import TYPE_CHECKING, Any, Dict, List
+
+from rich.console import Console
+from rich.rule import Rule
+from rich.syntax import Syntax
+
+from smolagents.models import MessageRole
+from smolagents.utils import AgentError, make_json_serializable
+
+
+if TYPE_CHECKING:
+    from smolagents.models import ChatMessage
+
+
+console = Console()
+
+
+@dataclass
+class Message:
+    role: MessageRole
+    content: str | list[dict]
+
+    def dict(self):
+        return asdict(self)
+
+
+@dataclass
+class ToolCall:
+    name: str
+    arguments: Any
+    id: str
+
+    def dict(self):
+        return {
+            "id": self.id,
+            "type": "function",
+            "function": {
+                "name": self.name,
+                "arguments": make_json_serializable(self.arguments),
+            },
+        }
+
+
+class AgentStepLog:
+    raw: Any  # This is a placeholder for the raw data that the agent logs
+
+    def dict(self):
+        return asdict(self)
+
+    def to_messages(self, **kwargs) -> List[Dict[str, Any]]:
+        raise NotImplementedError
```

Comment:
Wow, I think we nearly proposed the same at the same time! :open_mouth: 
See: https://github.com/albertvillanova/smolagents/commit/75ffbed2db311d9585bb36e4a4f45ed9df1b3c00#diff-e03ffeb1ffdc0c18d4a29eaec2694a39de7c9ce98647a9dcf5fc372a61d4d40aR83-R84

However, I have been thinking about it and if it could be implemented better... Not a clear idea yet, but let me think a bit more.